### ISSUE TITLE
Inconsistent and broken Auto import in VSCode

### ISSUE THREAD

**Issue by ackvf** [Microsoft/TypeScript#23014]
Hereby I confirm I did search and read dozen of auto import issues, but they were not related or outdated.

**TypeScript Version:**  2.7.2

**Search Terms:**  auto import relative absolute tsconfig baseUrl paths

**Code**
This is my project structure and code. Look for `Account`.
![image](https://user-images.githubusercontent.com/2313018/38132923-5375cd16-340d-11e8-828d-912d61b1d2c4.png)

```ts
//tsconfig.json
    "baseUrl": "./",
    "paths": {
      "~/*": ["src/*"]
    }
```

**Expected behavior:**

1. With `paths` turned **OFF** _(code commented)_
Expected is that I am offered with correct _relative_ paths
   1. for `index.ts` file
   `Auto import from './Account/Account' ; class Account` - _import is already correct_
   `Auto import from './routes' ; CONST ACCOUNT` - _correct_
   2. for `SignIn.tsx` file
   `Auto import from '../Account/Account' ; class Account` - _this import is missing_
   `Auto import from '../routes' ; CONST ACCOUNT` - _correct_
2. With `paths` turned **ON** _(code uncommented)_
Expected is that I am offered with correct _relative_ **AND** _absolute_ paths
   1. for `index.ts` file
   `Auto import from './Account/Account' ; class Account` - _missing_
   `Auto import from '~/pages/Account/Account' ; class  Account` - _correct_
   `Auto import from './routes' ; CONST ACCOUNT` - _missing_
   `Auto import from '~pages/routes' ; CONST ACCOUNT` - _correct_
   2. for `SignIn.tsx` file
   `Auto import from '../Account/Account' ; class Account` - _missing_
   `Auto import from '~/pages/Account/Account' ; class  Account` - _missing_
   `Auto import from '../routes' ; CONST ACCOUNT` - _missing_
   `Auto import from '~pages/routes' ; CONST ACCOUNT` - _correct_

**Actual behavior:**
_note that there are also some completely wrong import suggestions_
1.  `paths` **OFF**
    1. error1: path `'.'` is invalid - there is no localy available export
![import1](https://user-images.githubusercontent.com/2313018/38133599-3415150a-3410-11e8-9a60-cdc60581aead.png)
    2. error2: no class imports are available at all
![image](https://user-images.githubusercontent.com/2313018/38133694-9a68abaa-3410-11e8-9717-4058d6cead3b.png)
2.  `paths` **ON**
    1.  error3: path `'~/pages'` is invalid - there is no available export
![import2](https://user-images.githubusercontent.com/2313018/38134686-78063c5e-3414-11e8-8db8-cccaa97769f3.png)
    2. error4: no class imports are available at all
![image](https://user-images.githubusercontent.com/2313018/38134747-bd3bf840-3414-11e8-85a3-6fbffecc38d2.png)

note: Auto import in a root file `App.tsx` finds the class exports.
note2: There is still the wrong import from `src/pages` folder.
`paths` **OFF**
![import4](https://user-images.githubusercontent.com/2313018/38135120-ac24dad4-3416-11e8-8139-82130e49dfcc.png)
`paths` **ON**
![import3](https://user-images.githubusercontent.com/2313018/38135051-59a3d53a-3416-11e8-9d07-6250c7baede7.png)

So far I have counted 6 missing import suggestions and 4 wrong import suggestions. I am verry confused by the actual behaviour. It looks like I can import from any parent folder of importing module, but not from a sibling folder.
Restarting computer didn't help.

Am I missing something? What's wrong with my configuration?

**Related Issues:**
[Autoimport always imports relative to baseUrl #19920](https://github.com/Microsoft/TypeScript/issues/19920)

**Comment by andy-ms** [comment]
Could you try and reproduce this in a simple project with only a few files?

**Comment by ackvf** [comment]
I have actually just started this project. I removed all credentials and all references to firebase, here's the repo. https://github.com/ackvf/vscode-auto-import

**Comment by andy-ms** [comment]
> error1: path '.' is invalid - there is no localy available export

`.` is the current directory, and node will look in `./index.ts` which exports `Account`. That's a bit strange since it's the current file, but it seems to work.

> error2: no class imports are available at all

In `SignIn.tsx` with path mapping off, I can get an import `import { Account } from '..';` -- sometimes. It seems like sometimes vscode isn't asking for the completion details, so the completion shows up but there is no auto-import. @mjbvz 

> error3: path '~/pages' is invalid - there is no available export

It seems valid? `~/pages` is mapped to `src/pages`, which is imported from `src/pages/index.ts`, which has an `Account` export, same as with error1.

> error4: no class imports are available at all

As with error2, this is working for me half the time. The other half, vscode doesn't seem to be requesting completion details.

**Comment by ackvf** [comment]
@andy-ms I am sorry, I fixed all imports by hand before pushing to github. The `Account` import was not there by the time I reported the error. See screenshot above.

**Comment by andy-ms** [comment]
So you can't reproduce any issue any more?

I've filed an issue at Microsoft/vscode#47988 about the other problems I noticed.

**Comment by ackvf** [comment]
Actually the auto-import still doesn't work. There is another discovery.

![auto-fixes](https://user-images.githubusercontent.com/2313018/38856557-1353370a-4227-11e8-96ce-ea7b8e0c4746.png)

**Comment by andy-ms** [comment]
What's the bug in that image?

**Comment by ackvf** [comment]
The bug is that `ctrl+space` does not suggest what the light-bulb does.
`Import 'Navlink' from module "~/component/links/NavLink"` is missing.

**Comment by andy-ms** [comment]
Import fix is designed to give you a choice of all possible ways of importing something. Completions just chooses what it thinks is the best one.

**Comment by FredyC** [comment]
Um, is there some keyboard shortcut for opening that yellow bulb options? I don't even know how it's called, so it's hard to find some :) I've also run into many situations where import was not among the completions so I had to click with the mouse 🙄 to pick correct option. It's still faster then typing that import by hand for sure, but could be better :)

**Comment by andy-ms** [comment]
That screenshot shows that the completions are giving you one way of getting the identifier. We intentionally avoid adding many different completions for the same identifier.

**Comment by ackvf** [comment]
Right, but there are actually two sources for `NavLink`, one is `react-router-dom`, another is my own implementation. I believe it should be there, it's not like the same component available from multiple export paths.

**Comment by ackvf** [comment]
I will test it tomorrow. Thank you for your reply.

**Comment by ackvf** [comment]
I see! Actually I continued working on the project so the `NavLink` was indeed not in the example repo. But the problem is clearly visible in the gif. And fortunatelly-unfortunatelly I reproduced the issue in another commit in aforementioned repo. Try opening `components\Navigation\index.tsx` and import the missing `NavLink` from there on [line 20](https://github.com/ackvf/vscode-auto-import/blob/6eed5ae4bc1b94cd7e7d5dc9746e6303f9fd1f01/src/components/Navigation/index.tsx#L20).
